### PR TITLE
Make `flex` a `num?`

### LINKED ISSUE #22512: The flex property should have type double instead of int
Why is the `flex` property (e.g. in the `Flexible` widget) an int instead of a double?  If I for example want to use the golden ratio for relative sizing, I now have to use 161803398875 and 100000000000 as the (approximate) relative values.

### PR DESCRIPTION
**Pull request by ValentinVignal**
Fixes https://github.com/flutter/flutter/issues/22512

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by ValentinVignal** [comment]
My implementation has a breaking change.

Before: 
`flex: 0` would use the size of the child

After:
`flex: 0` would have a size 0 (for continuity)
`flex: null` would use the size of the child


Is it too big of a breaking change? If not, I can look at a data driven fix

**Comment by ValentinVignal** [comment]
Hello @goderbauer, do you have any feedback on this pull request?

**Comment by goderbauer** [comment]
> My implementation has a breaking change.

What's the motivation for this breaking change? Instead of sizing the children to 0 they should probably just be removed altogether from the tree?

**Comment by ValentinVignal** [comment]
It was for continuity. Ex: If I want my child to get smaller and smaller until it disappears:

```dart
AnimatedBuilder(
  animation: animation, // Animation going from 1 to 0 or 0 to 1
  builder: (context, child) {
    return Expanded(
      flex: animation.value,
      child: child,
    ),
  },
  child: child,
);
``` 


If felt if was making more sense

**Comment by ValentinVignal** [comment]
@goderbauer what are your thoughts on this? Should we discuss it somewhere?

**Comment by goderbauer** [comment]
I don't think that breaking change is worth it. We don't want to encourage people to keep widgets in the tree that are not actually visible/used as that can yield hard-to-debug performance problems.

**Comment by ValentinVignal** [comment]
@goderbauer I reverted it back to the default behavior in [revert: Remove breaking change when flex is zero](https://github.com/flutter/flutter/pull/147976/commits/cd49b1faf788343da798d9c90a40c7944baa3540)

File: packages/flutter/lib/src/rendering/flex.dart [hunk 1]
```diff
@@ -117,7 +117,7 @@ class FlexParentData extends ContainerBoxParentData<RenderBox> {
   /// non-zero, the amount of space the child's can occupy in the main axis is
   /// determined by dividing the free space (after placing the inflexible
   /// children) according to the flex factors of the flexible children.
-  int? flex;
+  num? flex;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/flex.dart:120]: Can you add a note to the docs about why this is typed `num` and that you'd typically pass in an `int` (or `double`)?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -5203,7 +5203,7 @@ class Flexible extends ParentDataWidget<FlexParentData> {
   /// non-zero, the amount of space the child can occupy in the main axis is
   /// determined by dividing the free space (after placing the inflexible
   /// children) according to the flex factors of the flexible children.
-  final int flex;
+  final num? flex;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:5206]: Same here.

File: packages/flutter/lib/src/rendering/flex.dart (hunk 1 above)
Comment (ValentinVignal) [reply packages/flutter/lib/src/rendering/flex.dart:124]: Sure, I updated it in [doc: Update the doc](https://github.com/flutter/flutter/pull/147976/commits/593490d596a6bfb006e657be5915606358d147ec)

**Comment by goderbauer** [comment]
@Hixie elsewhere I saw you raise some performance concerns about moving from num to double. Do you think that would be a concern here as well?

**Comment by Hixie** [comment]
You'd have to measure it, but in general, I would always assume that anything that requires more instructions is slower than code that requires fewer instructions.

Personally I don't see much benefit to changing this. Anything you can do with doubles you can already do with ints if you scale up all your values.

**Comment by ValentinVignal** [comment]
@Hixie @goderbauer I can measure it if you want. What kind of measurements/tests should I make?

**Comment by Hixie** [comment]
You'd probably want a microbenchmark comparing doing polymorphic math with `num` where you're passing in both doubles and ints, vs monomorphic math with `int`. I suspect you could just decompile a trivial example of this in godbolt and see that it's many more instructions, though.

**Comment by ValentinVignal** [comment]
@Hixie I've run this code on the dartpad:

<details>

<summary>Code</summary>

```dart
void intFunction(List<int> ints, int zero, int one) {
  final add = ints.fold(zero, (a, b) => a + b);
  final substract = ints.fold(zero, (a, b) => a - b);
  final multiply = ints.fold(one, (a, b) => a * b);
  final divide = ints.fold(one.toDouble(), (a, b) => a / b);
}

void doubleFunction(List<double> ints, double zero, double one) {
  final add = ints.fold(zero, (a, b) => a + b);
  final substract = ints.fold(zero, (a, b) => a - b);
  final multiply = ints.fold(one, (a, b) => a * b);
  final divide = ints.fold(one, (a, b) => a / b);
}

void numFunction(List<num> ints, num zero, num one) {
  final add = ints.fold(zero, (a, b) => a + b);
  final substract = ints.fold(zero, (a, b) => a - b);
  final multiply = ints.fold(one, (a, b) => a * b);
  final divide = ints.fold(one.toDouble(), (a, b) => a / b);
}

const size = 10000000;

void main() {
  // Ints
  
  final List<int> ints = List.generate(size, (index) => index + 1);
  final startInt = DateTime.now();

  intFunction(ints, 0, 1);

  final endInt = DateTime.now();
  final intDuration = endInt.difference(startInt);
  print('int: $intDuration');

  // Doubles

  final List<double> doubles = List.generate(size, (index) => index + 1.0);
  final startDouble = DateTime.now();

  doubleFunction(doubles, 0.0, 1.0);

  final endDouble = DateTime.now();
  final doubleDuration = endDouble.difference(startDouble);
  print('double: $doubleDuration');

  // Num ints
  
  final List<num> numInt = List.generate(size, (index) => index + 1);
  final startNumInt = DateTime.now();

  numFunction(numInt, 0, 1);

  final endNumInt = DateTime.now();
  final numIntDuration = endNumInt.difference(startNumInt);
  print('num int: $numIntDuration');

  // Num doubles
  
  final List<num> numDouble = List.generate(size, (index) => index + 1.0);
  final startNumDouble = DateTime.now();

  numFunction(numDouble, 0.0, 1.0);

  final endNumDouble = DateTime.now();
  final numDoubleDuration = endNumDouble.difference(startNumDouble);
  print('num double: $numDoubleDuration');
}
```

</details>

Here are the logs:

```console
int: 0:00:06.317000
double: 0:00:07.287000
num int: 0:00:07.894000
num double: 0:00:08.258000
```


I'm not too sure what kind of code you would expect me to test. Does this one make sense for you?

**Comment by Hixie** [comment]
Yeah, that's pretty good. Running it in dartpad will give you an idea of the performance for Dart compiled to JS. JS converts everything to doubles regardless, so I'd expect the numbers to be the same across the board, so it's interesting that even in that case we're seeing num be slower.

I tweaked your code a bit:

<details>

```dart
int intFunction(List<int> list, int zero, int one) {
  return list.fold<int>(zero, (int a, int b) => a + b) +
         list.fold<int>(zero, (int a, int b) => a - b) +
         list.fold<int>(one, (int a, int b) => a * b) +
         list.fold<int>(one, (int a, int b) => a ~/ b);
}

double doubleFunction(List<double> list, double zero, double one) {
  return list.fold<double>(zero, (double a, double b) => a + b) +
         list.fold<double>(zero, (double a, double b) => a - b) +
         list.fold<double>(one, (double a, double b) => a * b) +
         list.fold<double>(one, (double a, double b) => a / b);
}

num numFunction(List<num> list, num zero, num one) {
  return list.fold<num>(zero, (num a, num b) => a + b) +
         list.fold<num>(zero, (num a, num b) => a - b) +
         list.fold<num>(one, (num a, num b) => a * b) +
         list.fold<num>(one, (num a, num b) => a / b);
}

const int size = 100000000;

void main() {
  final Stopwatch stopwatch = Stopwatch();
  
  // Ints
  final List<int> ints = List<int>.generate(size, (int index) => index + 1);
  stopwatch..reset()..start();
  intFunction(ints, 0, 1);
  print('int,${stopwatch.elapsedMicroseconds}μs');

  // Doubles
  final List<double> doubles = List<double>.generate(size, (int index) => index + 1.0);
  stopwatch..reset()..start();
  doubleFunction(doubles, 0.0, 1.0);
  print('double,${stopwatch.elapsedMicroseconds}μs');

  // Num ints
  final List<num> numInt = List<num>.generate(size, (int index) => index + 1);
  stopwatch..reset()..start();
  numFunction(numInt, 0, 1);
  print('num int,${stopwatch.elapsedMicroseconds}μs');

  // Num doubles
  final List<num> numDouble = List<num>.generate(size, (int index) => index + 1.0);
  stopwatch..reset()..start();
  numFunction(numDouble, 0.0, 1.0);
  print('num double,${stopwatch.elapsedMicroseconds}μs');

  // Mixed num
  final List<num> numMixed = List<num>.generate(size, (int index) => index.isEven ? index : index + 1.0);
  stopwatch..reset()..start();
  numFunction(numMixed, 0, 1.0);
  print('num mixed,${stopwatch.elapsedMicroseconds}μs');
}
```
</details>

...and ran it using dart JIT a couple of times, using `flutter_test` (which should be the same but is slower for some reason) and with AOT compilation on Linux, and got the following results:

jit1, jit2, flutter_test and aot

This shows pretty clearly that `int` is much faster on that benchmark than `num` (especially in the case where we mix ints and doubles in the input). This is what I would expect -- `int` compiles to simple instructions, whereas `num` has to do a type check before every operation, introducing more branches, and more code in general. More code is generally slower than less code, all else being equal.

**Comment by Hixie** [comment]
Given that there's a know performance drop and you can more or less do anything you can with floats here using ints, I don't think the trade-off is especially compelling, so I would probably lean towards closing.

(I would generally recommend avoiding even barely-measurable performance drops much smaller than this, because at the end of the day, the speed of Flutter as a whole is a sum of all the performance drops. Ten thousand one-microsecond drops, and you've dropped ten milliseconds, even if each individual one couldn't be measured. When the entire budget of building a frame is 4ms, you very quickly suffer when you take on the performance hits, even if each one is small.)

**Comment by ValentinVignal** [comment]
Sure, I can close this PR then. Can we also close this issue to prevent people from spending time on this since we know we won't do it?

**Pull request closed without merging**